pkg/riff/commands/credential_delete_test.go

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package commands_test
+
+import (
+	"github.com/projectriff/riff/pkg/riff/commands"
+	"github.com/projectriff/riff/pkg/testing"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestCredentialDeleteCommand(t *testing.T) {
+	t.Parallel()
+
+	credentialName := "test-credential"
+	credentialAltName := "test-alt-credential"
+	defaultNamespace := "default"
+	credentialLabel := "projectriff.io/credential"
+
+	table := testing.CommandTable{
```

Comment:
A general observation about this style of test versus ginkgo is that the table is flat whereas a ginkgo test is structured as a tree which makes it easy to add test variations without much repetition. The tree structure also helps me, and hopefully others, to think through the cases and sub-cases of the test. When I want to work on a particular node of the ginkgo tree, I can focus it, add extra tests to that node, etc.

One particular use case is TDD. I add a failing test, focus on it, implement the code to make the test pass and then, if there's a problem with the test, I can immediately use the debugger to see what's going on (without hitting the same breakpoint courtesy of other tests). I can then make code changes and ultimately remove the focus (and if I forget to do this, the build will break).

Reply:
Your usage of Ginkgo is the cleanest I've seen. And I see how it can be easier to add an incremental test. That said, I find a hierarchal structure harder to understand unless you created it (this is true for anything with a custom spatial dimension that doesn't exist in the physical world). It's much more complicated to refactor since you need to have a deep understanding of how any particular state is shared between any other tests. For testing, I find being explicit to always be preferable to being dry. Denormalization is good in this case.

I do agree that being able to focus on a single record under test is handy. I added `Skip` and `Focus` bool fields to the table record struct. I'm still not sold on this focus being better than a filter on the `go test` command. Let me know what you think and we can keep or remove it.

Reply:
> Your usage of Ginkgo is the cleanest I've seen. And I see how it can be easier to add an incremental test. That said, I find a hierarchal structure harder to understand unless you created it (this is true for anything with a custom spatial dimension that doesn't exist in the physical world). It's much more complicated to refactor since you need to have a deep understanding of how any particular state is shared between any other tests. For testing, I find being explicit to always be preferable to being dry. Denormalization is good in this case.

Yes, I can see there are some advantages of using a "flatter" and more explicit testing approach, even if it is quite a bit more verbose. I'm personally happy to try other approaches so long as they are reasonably stable and well documented. But I would say that we need to agree to this as a team and take time to spread the testing work around so we can all become familiar with the new approach. It would be all too easy for a couple of us to create a silo here.

> 
> I do agree that being able to focus on a single record under test is handy. I added `Skip` and `Focus` bool fields to the table record struct. I'm still not sold on this focus being better than a filter on the `go test` command. Let me know what you think and we can keep or remove it.

I think the `Focus` field will be quicker to set than writing a filter, so I'd like to keep it and see how it works out.

Reply:
Maybe this work in progress should be pushed to a branch on this repo, so we can start to split work amongst us and spread the work that way (requires some coordination though, but I think code review is not necessarily enough to fully grasp the pros and cons).